### PR TITLE
Cleanly define the co-moving origin and add moving window tests

### PR DESCRIPTION
**Pull request by ikbuibui**
- Redefined the co-moving origin to make it more simple and clear
- Added moving window tests

While doing this, I ran into a problem with including the param file in the dimensional tests. The following changes were also made
- fix the cmake to properly include the dimensional test param files
- use simDim instead of TEST_DIM in all dimensional tests
- use target_include_directories instead of include_directories

### CODE REVIEW

File: include/picongpu/simulation/control/MovingWindow.hpp [hunk 1]
```diff
@@ -275,12 +277,26 @@ namespace picongpu
...
+            endSlidingOnStep = 0u;
+            computedConstants.reset();
+            firstCallEndSlideOnStep = true;
         }
 
         /**
          * Set step where the simulation stops the moving window
          *
          * @param step 0 means no sliding window, else sliding is enabled until step is reached.
+         * Negative numbers are the way to use this if if don't want the sliding to end (practically) -1 sets it to the
```
Comment (chillenzer) [inline include/picongpu/simulation/control/MovingWindow.hpp]:
```suggestion
         * Negative numbers are the way to use this if don't want the sliding to end (practically) -1 sets it to the
```

File: share/picongpu/unit/dimensional_tests/MovingWindow.cpp
Comment (chillenzer) [inline share/picongpu/unit/dimensional_tests/MovingWindow.cpp:1]: :tada: Tests!

File: share/picongpu/unit/dimensional_tests/MovingWindow.cpp [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+#include <catch2/catch_test_macros.hpp>
+#include <catch2/generators/catch_generators.hpp>
+#include <picongpu/defines.hpp>
+#include <picongpu/simulation/control/MovingWindow.hpp>
+
+using namespace picongpu;
+
+//! Helper to setup the PMacc environment
+using TestFixture = pmacc::test::PMaccFixture<simDim>;
+static TestFixture pmaccFixture;
```
Comment (chillenzer) [inline share/picongpu/unit/dimensional_tests/MovingWindow.cpp]: Not just this?
```suggestion
static pmacc::test::PMaccFixture<simDim> pmaccFixture;
```

File: share/picongpu/unit/dimensional_tests/MovingWindow.cpp [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    MovingWindowTestFixture() : window(MovingWindow::getInstance())
+    {
+        window.resetForTesting();
+    }
+
+    ~MovingWindowTestFixture()
+    {
+        window.resetForTesting();
+    }
+};
```
Comment (chillenzer) [inline share/picongpu/unit/dimensional_tests/MovingWindow.cpp:50]: That is a clear symptom of badly testable design: There's nothing about a moving window that needs to be a singleton. We should split the implementation of the moving window into a class providing the functionality that you can easily just instantiate and a form of storage that can happily still be a singleton. Might be beyond the scope of this PR though, I'm aware of that.

File: share/picongpu/unit/dimensional_tests/MovingWindow.cpp [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+TEST_CASE("unit::MovingWindow_origin", "[movingWindow test]")
+{
+    MovingWindowTestFixture fixture;
+    auto& movingWindow = fixture.window;
+
+    pmaccFixture.initGrids(
+        pmacc::DataSpace<3>{500, 1000, 500}.shrink<simDim>(),
+        pmacc::DataSpace<3>{100, 100, 100}.shrink<simDim>(),
+        pmacc::DataSpace<3>{0, 0, 0}.shrink<simDim>());
```
Comment (chillenzer) [inline share/picongpu/unit/dimensional_tests/MovingWindow.cpp]: Could we use all different numbers here? It's a bit of mental exercise, I'm aware, but it
1) exhibits bugs that might be hidden due to accidental equality
2) makes it easier to identify the problem in case there's a failure.

Of course, argument (1) works in reverse as well, so the best would be to use a Catch2 generator expression to generate tests with equal and different arguments and make the whole independent of the precise values.

File: share/picongpu/unit/dimensional_tests/MovingWindow.cpp [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+    auto const globalWindowSizeInMoveDirection = 900;
+
+    auto const c = static_cast<float_64>(sim.pic.getSpeedOfLight());
+    auto const dt = static_cast<float_64>(sim.pic.getDt());
+    // should both be 1 for correctness of the tests
+    assert(c == 1. && dt == 1.);
```
Comment (chillenzer) [inline share/picongpu/unit/dimensional_tests/MovingWindow.cpp]:
```suggestion
    REQUIRE(c == 1.);
    REQURE(dt == 1.);
```

File: share/picongpu/unit/dimensional_tests/MovingWindow.cpp [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+
+    auto const cellSize = sim.pic.getCellSize();
+
+    constexpr auto moveDirection = MovingWindow::moveDirection;
+    auto const cellSizeInMoveDirection = static_cast<float_64>(cellSize[moveDirection]);
+    auto const cellsPerStep = 1. / cellSizeInMoveDirection;
```
Comment (chillenzer) [inline share/picongpu/unit/dimensional_tests/MovingWindow.cpp]: I would probably factor out this whole initialisation into an own function. The reader quickly skimming this to find out about a particular behaviour doesn't need to bother how precisely `cellsPerStep` was computed. In fact, they don't even need to see `c` and `dt` at all.
```suggestion
    auto [globalWindowSizeInMoveDirection, ...] = init();
```

File: share/picongpu/unit/dimensional_tests/MovingWindow.cpp [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+
+    SECTION("Window disabled")
+    {
+        movingWindow.setEndSlideOnStep(0);
+        REQUIRE_FALSE(movingWindow.isEnabled());
+
+        auto origin = movingWindow.getMovingWindowOriginPositionCells(100);
```
Comment (chillenzer) [inline share/picongpu/unit/dimensional_tests/MovingWindow.cpp]: Don't use magic numbers...
```suggestion
        auto origin = movingWindow.getMovingWindowOriginPositionCells(arbitrary_step);
```

File: share/picongpu/unit/dimensional_tests/MovingWindow.cpp [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+
+        // Calculate firstMoveStep again for the given movePoint.
+        auto const virtualParticleInitialStartCell = static_cast<uint32_t>(
+            std::ceil(static_cast<float_64>(globalWindowSizeInMoveDirection) * (1.0 - movePoint)));
+
+        auto const wayToFirstMove
+            = static_cast<float_64>(globalWindowSizeInMoveDirection - virtualParticleInitialStartCell)
+              * cellSizeInMoveDirection;
+
+        auto const firstMoveStep = static_cast<int32_t>(std::ceil(wayToFirstMove / 1.)) - 1;
```
Comment (chillenzer) [inline share/picongpu/unit/dimensional_tests/MovingWindow.cpp]: Did you know that you can nest `SECTION`s? Catch2 will generate a tree of sections and subsections and every leave of that tree will define a separate run in which all other branches are skipped. That would significantly reduce code duplication here. Roughly:

```C++
SECTION("Window is enabled") {
    float_64 const movePoint = 0.5;
    // ...
    
    SECTION("but not yet moving.") {
        // ...
        auto const origin = movingWindow.getMovingWindowOriginPositionCells(firstMoveStep - 1);
        // ...
    }
    
    SECTION("and moving.") {
        // ...
        auto const origin = movingWindow.getMovingWindowOriginPositionCells(firstMoveStep + 1);
        // ...
    }
}
```
will generate (functionally) the same code that you wrote with duplication.

Btw. do you have a test precisely at the `firstMoveStep`? What weird edge cases can we run into there?

File: share/picongpu/unit/dimensional_tests/MovingWindow.cpp
Comment (chillenzer) [inline share/picongpu/unit/dimensional_tests/MovingWindow.cpp:1]: On a more serious note, I think these tests are very valuable. Although from what I hear, the moving window is a particularly fiddly thing and I'd strongly expect that these fill up quickly with regression tests and edge cases. But nevertheless: Great to have a start! Thanks!

File: share/picongpu/unit/dimensional_tests/TWTSTight.cpp [hunk 9]
```diff
@@ -44,13 +44,13 @@
...
-//! Helper to setup the PMacc environment
-using TestFixture = pmacc::test::PMaccFixture<TEST_DIM>;
-static TestFixture fixture;
-
 using namespace picongpu;
 using namespace pmacc;
 
+//! Helper to setup the PMacc environment
+using TestFixture = pmacc::test::PMaccFixture<simDim>;
+static TestFixture fixture;
```
Comment (chillenzer) [inline share/picongpu/unit/dimensional_tests/TWTSTight.cpp]: Same as before, I presume we don't actually use that typedef anywhere else? But, that's cosmetics of course. Only act on these if you happen to find yourself pushing changes anyways.

File: include/picongpu/simulation/control/MovingWindow.hpp [hunk 10]
```diff
@@ -381,6 +396,11 @@ namespace picongpu
 
         /**
          * Get the moving window origin position in cell, not discretized to the cell grid
+         * This position is the total origin (0,0,0) until we get to the first move step
+         * On the first move step the position is equal to the moving window move location plus the distance moved with
+         * c in one timestep. This ensures that this comoving particle is always inside the region of the moving window
+         * After this, the particle continues moving with c, until we reach the step where the window stops
+         * moving, Then, the particle stops where ever it is (maybe in the middle of a cell)
```
Comment (chillenzer) [inline include/picongpu/simulation/control/MovingWindow.hpp:403]: The function you're commenting on here explicitly has "Cells" in the name (`getMovingWindowOriginPositionCells`). I would probably be gullible enough to be tricked into believing that this would return a vector of integers. Should we adjust this? One could even consider generalising this function to take the same (compile-time) arguments as `DomainInfo` does (`CELL`/`SUB_CELL` precision and maybe even the others). Of course, we should only do that if we see a clear usecase for this.

File: include/picongpu/simulation/control/MovingWindow.hpp [hunk 11]
```diff
@@ -394,23 +414,24 @@ namespace picongpu
...
-            {
-                auto cellsPerStep = constants.deltaWayPerStep / constants.cellSizeInMoveDirection;
+            auto cellsPerStep = constants.deltaWayPerStep / constants.cellSizeInMoveDirection;
 
-                SubGrid<simDim> const& subGrid = Environment<simDim>::get().SubGrid();
-                float_64 const virtualParticleCellsPassed = cellsPerStep * static_cast<float_64>(currentStep);
-                float_64 const moveLocationInCells
-                    = constants.moveRatio
-                      * (subGrid.getGlobalDomain().size[moveDirection] - constants.gpuNumberOfCellsInMoveDirection);
+            auto firstStep = constants.firstMoveStep > 0 ? constants.firstMoveStep : 0;
```
Comment (chillenzer) [inline include/picongpu/simulation/control/MovingWindow.hpp:428]: Why would this ever be necessary? Can this be negative?

**Review by chillenzer — CHANGES_REQUESTED** [review]
It's not exactly blocking. Just forcing some discussion on this before you get the approval and merge.

File: share/picongpu/unit/dimensional_tests/MovingWindow.cpp (hunk 8 above)
Comment (ikbuibui) [reply share/picongpu/unit/dimensional_tests/MovingWindow.cpp]: I have combined the sections as suggested.
I have only sort of tested precisely at `firstMoveStep`. I have the test where the window starts moving at step zero, which is also the `firstMoveStep`. 
I had to be careful how to define the co-moving particle to deal with some edge cases. 
By being careful I mean that I needed to realize that the moving window can possibly already move on `firstMoveStep`, depending on the cell size and the move point. 
This is why the co-moving particle has an offset of `c*dt` to make sure for all cases that the co-moving particle is inside the moving window, since if the first move happens at first move step the moving window location would jump by one cell and this might leave the co-moving particle behind. Now the co-moving particle is always inside the first cell of the moving window, with the effect that particle positions relative to this origin might be negative..
I will add a test for this actually, making sure the particle is always in the first cell.

File: share/picongpu/unit/dimensional_tests/MovingWindow.cpp (hunk 2 above)
Comment (ikbuibui) [reply share/picongpu/unit/dimensional_tests/MovingWindow.cpp]: Done

File: share/picongpu/unit/dimensional_tests/MovingWindow.cpp (hunk 5 above)
Comment (ikbuibui) [reply share/picongpu/unit/dimensional_tests/MovingWindow.cpp]: Done.

File: share/picongpu/unit/dimensional_tests/MovingWindow.cpp (hunk 7 above)
Comment (ikbuibui) [reply share/picongpu/unit/dimensional_tests/MovingWindow.cpp]: Done,

File: share/picongpu/unit/dimensional_tests/TWTSTight.cpp (hunk 9 above)
Comment (ikbuibui) [reply share/picongpu/unit/dimensional_tests/TWTSTight.cpp]: Done

File: share/picongpu/unit/dimensional_tests/MovingWindow.cpp (hunk 6 above)
Comment (ikbuibui) [reply share/picongpu/unit/dimensional_tests/MovingWindow.cpp]: I see your point but am not particularly enthusiastic about this, since I think if someone wants to start reasoning about the logic of the tests they might care if these values are calculated correctly. Also a long structured binding isn't exactly pretty either. I do agree `c` and `dt` can be completely hidden, but I am for keeping it as it is to not add indirection. 
So unless you really want me to change this, I am going to leave it

File: include/picongpu/simulation/control/MovingWindow.hpp (hunk 11 above)
Comment (ikbuibui) [reply include/picongpu/simulation/control/MovingWindow.hpp:428]: Yes, it is equal to `-1` if `movePoint` is set to `0`.

**Review by ikbuibui — COMMENTED** [review]
Pushed a new commit with more tests, and applied most reviewer comments

File: include/picongpu/simulation/control/MovingWindow.hpp (hunk 1 above)
Comment (ikbuibui) [reply include/picongpu/simulation/control/MovingWindow.hpp]: Fixed

File: share/picongpu/unit/dimensional_tests/MovingWindow.cpp (hunk 4 above)
Comment (ikbuibui) [reply share/picongpu/unit/dimensional_tests/MovingWindow.cpp]: I have changed the number to not be identical anymore, however I decided not to put in the effort to test the case where they are equal

**Comment by PrometheusPi** [comment]
Started from a discussion weeks ago with @ikbuibui, I set up a test of the co-moving position using the binning plugin.
Below you see the result for the current dev (minor jumps) vs. this PR, no jumps. For reference, the y-cell-size is 0.04µm, which is also the binning width of the binning plugin.

grafik

File: include/picongpu/simulation/control/MovingWindow.hpp (hunk 10 above)
Comment (chillenzer) [reply include/picongpu/simulation/control/MovingWindow.hpp:403]: Offline discussion: The name is correct in that it refers to the units returned. The number will be a (floating point, so potentially fractional) multiple of the cell size.

File: share/picongpu/unit/dimensional_tests/MovingWindow.cpp [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+        // Manually calculate when the window should start moving to verify pre-movement phase.
+        auto const virtualParticleInitialStartCell = static_cast<uint32_t>(
+            std::ceil(static_cast<float_64>(globalWindowSizeInMoveDirection) * (1.0 - movePoint)));
+
+        // Calculate firstMoveStep again for the given movePoint.
+        auto const wayToFirstMove
+            = static_cast<float_64>(globalWindowSizeInMoveDirection - virtualParticleInitialStartCell)
+              * cellSizeInMoveDirection;
+
+        auto const firstMoveStep = static_cast<int32_t>(std::ceil(wayToFirstMove / 1.)) - 1;
```
Comment (chillenzer) [inline share/picongpu/unit/dimensional_tests/MovingWindow.cpp:106]: Now I'm curious what horrible debugging resulted in `std::ceil(wayToFirstMove / 1.)`? Or could this possibly be a unnamed magic number?

**Review by chillenzer — APPROVED** [review]

File: share/picongpu/unit/dimensional_tests/MovingWindow.cpp (hunk 12 above)
Comment (ikbuibui) [reply share/picongpu/unit/dimensional_tests/MovingWindow.cpp:106]: Oh actually this is only to find the integral time step where the window starts moving. It is just `ceil(time_taken) - 1`. It isn't `floor(time_taken)` because of implementation details of the moving window which instantly starts moving.
Maybe the `1.` is the source of confusion, which is `c*dt`

Comment (chillenzer) [reply share/picongpu/unit/dimensional_tests/MovingWindow.cpp:106]: Which is a magic constant!

Comment (chillenzer) [reply share/picongpu/unit/dimensional_tests/MovingWindow.cpp:106]: Why not `c*dt`?

Comment (BrianMarre) [reply share/picongpu/unit/dimensional_tests/MovingWindow.cpp:106]: Strictly speaking PIConGPU does not guarantee `c*dt == 1` that is just a convention in our typical param files

Comment (ikbuibui) [reply share/picongpu/unit/dimensional_tests/MovingWindow.cpp:106]: Yes, which is why the tests `REQUIRE` that both are actually one. I did start with `c*dt` everywhere, but IMO it started to become hard to parse through, so I opted for this.

Comment (chillenzer) [reply share/picongpu/unit/dimensional_tests/MovingWindow.cpp:106]: Nope, I still think that a named constant is the way to go. Just `c * dt` is not a good abstraction. I'd rather call it `characteristic_length` or `shift_length` or something... (those are admittedly not great but you get the idea).

Comment (ikbuibui) [reply share/picongpu/unit/dimensional_tests/MovingWindow.cpp:106]: Fair, I'll refactor the tests soon
